### ISSUE TITLE
default Debian 8 image--heavy-handed DNS config (thoughts, please)

### ISSUE THREAD

**Issue by gladiatr72** [kubernetes/kops#1095]
### TL;DR

It would be fabulous if kops were (in the context of the default Debian image) augment the local node ca-certificate store with the cluster CA certificate and reconfigure the node DNS  to point to the cluster DNS IP for local resolution--also to include `svc.cluster.local` as a search domain.

---

I have been using kops relatively successfully for the last few months.  Bravo! 🎉 

One thing that I had to spend time fighting with this week is the lack of attention to the local node's DNS configuration.  I understand that the k8s team's philosophy on the topic is: Every distribution is different--it is beyond the scope we have set for ourselves to try to deal with all of the various and sundry ways DNS works everywhere.

What is the kops team stance on this issue?  I am currently working with a single-az kops cluster on AWS (I haven't figured out a good way to deal with using persistent storage on multi-az clusters)  

  * Goal: define a solid pattern for self-contained CI/CD pipeline: ✅ 
     * jenkins pipeline ✅ 
        * spawn slave image ✅ 
        * sync source code ✅ 
        * build ✅ 
        * test ✅ 
        * create docker image ✅ 
        * push docker image to local registry ✅ 
        * deploy image to environment-labeled k8s node 👿 

> Note: I have made manual changes to my nodes for the purpose of getting on with my intended tasks

One of the details I have kept in mind is wanting to avoid reconfiguring docker (everywhere) to side-step the SSL issue--trying to embrace the whole _cattle not pets_ philosophy even unto the node layer! 😄   

So I've come up with a mechanism to generate a proper server certificate that encompasses `kube-repository.kube-system` and its alt names that include `svc|svc.cluster|svc.cluster.local`.  What doesn't work is to include the registry container's IP address--because who knows what that'll be, right?

But for This last bit to work, two things need to happen:

1. the node needs to be able to resolve cluster DNS host names.  
2. the node needs to recognize at least the cluster CA certificate as a trusted signer or (ideally) a user-specified CA certificate

With regards to DNS, whoever built this image has an obvious distain for anyone that might want to actually use it.  I tried superceding the DNS configuration via `dhclient.conf` (didn't work). I even attempted to create another resolver configuration file (`/etc/resolv.conf_kube`) for `kubelet` to point to via its `--resolv-conf` flag thinking that perhaps somehow its resolution visibility might get passed down to any docker processes it might spawn.

Anyway, I sort of resolved this by reconfiguring `/etc/resolv.conf` and setting its `immutable` flag.  

> Note: `dnsmasq` is *not* a solution in this context as whatever mechanism is clobbering changes (almost instantly) to /etc/resolv.conf is not subverted by dnsmasq's presence.

I look forward to reading your thoughts/ideas.

Thanks for reading!

-S

**Comment by chrislovecnm** [comment]
So where do you end up with this. We have been knee deep in a release, so apologize for the late response. What do you recommend? Also we need to get this over to the repo where the community image is maintained.

**Comment by gladiatr72** [comment]
My existing kube workers work as I need to for now. (on a deadline so I haven't had the oppotunity to come up with a satisfactory solution yet)

The big logic hole: the worker host must have access to the kube-dns service before setting the host to resolve from the cluster DNS service. This pertains to a new host as well as a host that has been cordoned -> uncordoned--it looks like `protokube` wants to pull new images (or I have great timing)--without DNS service this understandably fails.

Perhaps as part of the kops kublet/protokube/worker installation sequence:

  * is this host a worker
     *  verify TLS connection to API host
     *  is there a kube-proxy of the correct vintage in the docker cache
         * is there potential for image gc to sweep up the k8s images? I've still got `kube-proxy:1.4.6` in my workers' cache from 4 weeks ago; however, I have not put any sort of stress on these hosts' disk space
  
  * is the resolver IP a host on the k8s cluster network?
     * no: continue to bootstrap
     * yes: attempt resolving `kubernetes.srv` against this address
        * if it works: yay!
        * if it doesn't: restore the provider provided resolver IP and continue to bootstrap

As far as the local CA certificate is concerned, I discovered/learned-of `update-ca-certificates`.  
If you drop your CA cert into `/usr/share/ca-certificates/.` , add the CA cert's relative path to `/etc/ca-certificates.conf` and execute the update script, the Right Thing Is Done.  If you just try to add it to `/etc/ssl/certs/ca-certificates.crt` it will get overwritten.

**Comment by justinsb** [comment]
Didn't get to this in 1.5.0 (sorry!), but I'm marking it P1 for 1.5.1
